src/evolve.ts

```diff
@@ -0,0 +1,230 @@
+ * Creates return type from the type of arguments of `evolve`.
+ */
+type Evolve<T, E> = E extends AFunction
+  ? ReturnType<E>
+  : T extends object
+    ? {
+        [K in Keys<T, E> as GetValueByKey<T, K> extends never
+          ? never
+          : K]: Evolve<
+          GetValueByKey<T, K>,
+          GetValueByKey<E, K, GetValueByKey<T, K>>
+        >;
+      }
+    : T;
+
+/**
+ * Creates a new object by recursively evolving a shallow copy of `object`,
+ * according to the `transformation` functions. All non-primitive properties
+ * are copied by reference.
+ *
+ * A `transformation` function will not be invoked if its corresponding key
+ * does not exist in the evolved object.
+
+ * @param object object or array whose value at some path is applied to
+ * the corresponding function that is defined in `transformations` at the same path.
+ * @param transformations it is object or array in which the functions at some path is applied to
+ * the corresponding value of `object` at the same path.
+ * @signature
+ *    R.evolve(object, transformations)
+ * @example
+ *    const transf = {
+ *      count: add(1),
+ *      data: { elapsed: add(1), remaining: add(-1) },
+ *    };
+ *    const object = {
+ *      id: 10,
+ *      count: 10,
+ *      data: { elapsed: 100, remaining: 1400 },
+ *    };
+ *    evolve(object, transf)
+ *    // => {
+ *    //   id: 10,
+ *    //   count: 11,
+ *    //   data: { elapsed: 101, remaining: 1399 },
+ *    // }
+ * @dataFirst
+ * @category Object
+ */
+export function evolve<T extends EvolveTargetObject<E>, E>(
+  object: T,
+  transformations: E
+): Evolve<T, E>;
+
+/**
+ * Creates a new object by recursively evolving a shallow copy of `object`,
+ * according to the `transformation` functions. All non-primitive properties
+ * are copied by reference.
+ *
+ * A `transformation` function will not be invoked if its corresponding key
+ * does not exist in the evolved object.
+
+ * @param object object or array whose value at some path is applied to
+ * the corresponding function that is defined in `transformations` at the same path.
+ * @param transformations it is object or array in which the functions at some path is applied to
+ * the corresponding value of `object` at the same path.
+ * @signature
+ *    R.evolve(transf)(object)
+ * @example
+ *    const transf = {
+ *      count: add(1),
+ *      data: { elapsed: add(1), remaining: add(-1) },
+ *    };
+ *    const object = {
+ *      id: 10,
+ *      count: 10,
+ *      data: { elapsed: 100, remaining: 1400 },
+ *    };
+ *    R.pipe(object, R.evolve(transf))
+ *    // => {
+ *    //   id: 10,
+ *    //   count: 11,
+ *    //   data: { elapsed: 101, remaining: 1399 },
+ *    // }
+ * @dataLast
+ * @category Object
+ */
+export function evolve<E>(
+  transformations: E
+): <T extends EvolveTargetObject<E>>(object: T) => Evolve<T, E>;
+
+export function evolve() {
+  return purry(_evolve, arguments);
+}
+
+function _evolve(object: any, transformations: any) {
```

Comment:
I implemented it using the approach of using types as much as possible, and the result is shown [here](https://github.com/remeda/remeda/blob/541bb47603431805696ee7844e7089af7da481e7/src/evolve.ts#L248~L280).
Notations such as `result[key]` are strongly restricted by the result type and key type, so to avoid this, I separated array processing and object processing.
However, in the end, as described in [the comments here ❷](https://github.com/remeda/remeda/pull/454#issuecomment-1940654791), I decided not to handle arrays.